File: backend/app/routers/conversations.py
from __future__ import annotations
from fastapi import APIRouter, Depends
from supabase import Client

from app.deps import get_db, require_user
from app.services.conversation_service import ConversationService

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.get("/list")
def list_conversations(user=Depends(require_user), db: Client = Depends(get_db)):
    items = ConversationService(db).list_conversations(user.id)
    return {"ok": True, "data": {"conversations": items}}


@router.get("")
def get_history(scope: str = "global", scope_id: str | None = None, limit: int = 20,
                user=Depends(require_user), db: Client = Depends(get_db)):
    msgs = ConversationService(db).get_history(user.id, scope, scope_id, limit)
    data = [{"id": m.id, "role": m.role, "text": m.text, "at": m.at} for m in msgs]
    return {"ok": True, "data": {"messages": data}}


@router.delete("")
def delete_conversation(scope: str = "global", scope_id: str | None = None,
                        user=Depends(require_user), db: Client = Depends(get_db)):
    """Permanently delete the current user's conversation for this scope
    (and all of its messages via CASCADE)."""
    deleted = ConversationService(db).delete_conversation(user.id, scope, scope_id)
    return {"ok": True, "data": {"deleted": deleted}}


@router.post("/search")
def search(body: dict, user=Depends(require_user), db: Client = Depends(get_db)):
    query = body.get("query", "")
    scope = body.get("scope", "global")
    scope_id = body.get("scope_id")
    limit = body.get("limit", 20)
    msgs = ConversationService(db).search(user.id, query, scope, scope_id, limit)
    data = [{"id": m.id, "role": m.role, "text": m.text, "at": m.at} for m in msgs]
    return {"ok": True, "data": {"messages": data}}

File: frontend/lib/api/conversations.ts
import { apiGet, apiPost, apiDelete, ApiResult } from "./client";

export interface ConvMessage {
  id: string;
  role: "user" | "assistant" | "tool";
  text: string;
  at: string;
}

export interface ConvHistory {
  messages: ConvMessage[];
}

export interface ConversationSummary {
  id: string;
  scope: "global" | "plant" | "bud" | "calendar";
  scope_id: string | null;
  message_count: number;
  last_message: string | null;
  last_role: "user" | "assistant" | null;
  updated_at: string;
  created_at: string;
}

export interface ConversationList {
  conversations: ConversationSummary[];
}

export function listConversations(): Promise<ApiResult<ConversationList>> {
  return apiGet<ConversationList>("/conversations/list");
}

export function getHistory(
  scope = "global",
  scopeId?: string,
  limit = 40
): Promise<ApiResult<ConvHistory>> {
  const params = new URLSearchParams({ scope, limit: String(limit) });
  if (scopeId) params.set("scope_id", scopeId);
  return apiGet<ConvHistory>(`/conversations?${params}`);
}

export function deleteConversation(
  scope = "global",
  scopeId?: string
): Promise<ApiResult<{ deleted: number }>> {
  const params = new URLSearchParams({ scope });
  if (scopeId) params.set("scope_id", scopeId);
  return apiDelete<{ deleted: number }>(`/conversations?${params}`);
}

export function searchConversation(
  query: string,
  scope = "global",
  scopeId?: string,
  limit = 20
): Promise<ApiResult<ConvHistory>> {
  return apiPost<ConvHistory>("/conversations/search", { query, scope, scope_id: scopeId, limit });
}
